tools/slicec-cs/src/generators/struct_generator.rs

```diff
@@ -128,3 +121,55 @@ pub fn generate_struct(struct_def: &Struct) -> CodeBlock {
 
     builder.build()
 }
+
+pub fn generate_encoding_blocks(
+    fields: &[&Field],
+    encodings: SupportedEncodings,
+    encoding_fn: fn(&[&Field], Encoding) -> CodeBlock,
+    encoding_source: &'static str,
+) -> CodeBlock {
+    match encodings[..] {
+        [] => unreachable!("No supported encodings"),
+        [encoding] => encoding_fn(fields, encoding),
+        _ => {
+            let mut slice1_block = encoding_fn(fields, Encoding::Slice1);
+            let mut slice2_block = encoding_fn(fields, Encoding::Slice2);
+
+            // Only write one encoding block if `slice1_block` and `slice2_block` are the same.
+            if slice1_block.to_string() == slice2_block.to_string() {
+                return slice2_block;
+            }
+
+            if slice1_block.is_empty() && !slice2_block.is_empty() {
+                format!(
+                    "\
+if ({encoding_source}.Encoding != SliceEncoding.Slice1) // Slice2 only
+{{
+{slice2_block}
+}}
+",
+                    slice2_block = slice2_block.indent(),
+                )
+                .into()
+            } else if !slice1_block.is_empty() && !slice2_block.is_empty() {
+                format!(
+                    "\
+if ({encoding_source}.Encoding == SliceEncoding.Slice1)
+{{
+{slice1_block}
+}}
+else // Slice2
+{{
+{slice2_block}
+}}
+",
+                    slice1_block = slice1_block.indent(),
+                    slice2_block = slice2_block.indent(),
+                )
+                .into()
+            } else {
+                unreachable!("it's impossible to have an empty Slice2 block with a non empty Slice1 block");
+            }
```

Comment:
```suggestion
            match (slice1_block.is_empty(), slice2_block.is_empty()) {
                // Both slices are equal
                (false, false) if slice1_block.to_string() == slice2_block.to_string() => slice2_block,

                // Slice1 is empty, Slice2 is not
                (true, false) => format!(
                    "\
if ({encoding_source}.Encoding != SliceEncoding.Slice1) // Slice2 only
{{
{slice2_block}
}}
                ",
                    slice2_block = slice2_block.indent(),
                )
                .into(),

                // Neither Slice1 nor Slice2 is empty
                (false, false) => format!(
                    "\
if ({encoding_source}.Encoding == SliceEncoding.Slice1)
{{
{slice1_block}
}}
else // Slice2
{{
{slice2_block}
}}
                ",
                    slice1_block = slice1_block.indent(),
                    slice2_block = slice2_block.indent(),
                )
                .into(),

                // Other cases (Slice1 is not empty, Slice2 is)
                // or both are empty - which seems to be treated as an error in the original code.
                _ => unreachable!("it's impossible to have an empty Slice2 block

```

I think this would be a bit clearer of a way to write this. Following all of the ifs can be a bit trickY! (Also this suggestion probably needs additional formatting - was just to communicate the idea)